### PR TITLE
Collect custom attributes and tags

### PR DESCRIPTION
**Pull request by rmontenegroo**
Hi again!

I've been working around issue #106. 

> Unfortunately, I was not able to keep only using pyVmomi and I had to use Vmware Python SDK as well so as to collect tags.

That is in the past. I changed every code snippet using vmware SDK to direct requests to the vsphere rest api. I hope that is ok.

Please take a look.

### CODE REVIEW

File: vmware_exporter/helpers.py [hunk 1]
```diff
@@ -1,8 +1,20 @@
...
 from pyVmomi import vmodl
 
 
+def get_unverified_session():
+    """
+    create session to connect to vmware API
+    from Vmware Samples
+    """
+    session = requests.session()
+    session.verify = False
```
Comment (pryorda) [inline vmware_exporter/helpers.py:13]: Can you make this a defaulted variable?

File: vmware_exporter/helpers.py [hunk 2]
```diff
@@ -15,6 +27,19 @@ def batch_fetch_properties(content, obj_type, properties):
         recursive=True
     )
 
+    """
+        Gathering all custom attibutes names are stored as key (integer) in CustomFieldsManager
+        We do not want those keys, but the names. So here the names and keys are gathered to
+        be translated later
+    """
+    allCustomAttributesNames = dict(
```
Comment (pryorda) [inline vmware_exporter/helpers.py:35]: Can you add an example of what this looks like as a metric?

File: vmware_exporter/vmware_exporter.py [hunk 3]
```diff
@@ -32,172 +44,223 @@
...
+        self._vmsCustomAttributes = {}
+        self._hostsCustomAttributes = {}
+        self._datastoresCustomAttributes = {}
+
+        # Tags
+        # flag to wheter fetch tags or not
+        self.fetch_tags = fetch_tags
+
+        # label names and ammount will be needed later to insert labels from custom attributes
+        self._labelNames = {
```
Comment (pryorda) [inline vmware_exporter/vmware_exporter.py:85]: I like what you did here.

**Comment by Jc2k** [comment]
Full disclosure - the project i was using vmware_exporter for is now k8s bare metal w/ kata, so I don't have a test cluster for this and my head is out of the game.

It's a big diff and haven't given't it a thorough review but seems like a reasonable change. I'd probably want to test it against a cluster before merging it, if I still had one.

My biggest concern would be vmware's code, TBH. It's really grim that vmware can't manage to put an wheel on pypi for their library in something like 2.5 years (see https://github.com/vmware/vsphere-automation-sdk-python/issues/38). My gut reaction is if they can't manage that, then what other python basics are they not doing? Anyway I poked around and apparently they feel like they can't list it on pypi cause its not open source enough. The community seems dubious about the validity of this, but i can't see them give any other rationale for not doing it. If they think it's not open source enough to list on pypi maybe we need to consider whether it's an appropriate dependency for vmware_exporter.

If you don't think the licencing is an issue, then the next thing would be pinning the github dependency to a known sha so if they push a regression or deprecation we don't pick it up immediately by surprise.

I'm also not sure about referencing a github commit from setup.py (requirements.txt gets hoovered into setup.py). I generally avoid doing depending on unreleased stuff, so I have no idea if it actually works. So we should probably test that it can generate valid wheels still.

**Comment by rmontenegroo** [comment]
I've been testing on a DC with more than 6k vms,  60 hosts, 8 clusters and
around 260 ds. Runs pretty good to me. I can't see much difference.

Em sáb, 25 de abr de 2020 12:44, Daniel Pryor <notifications@github.com>
escreveu:

**Comment by rmontenegroo** [comment]
About the license:
https://github.com/vmware/vsphere-automation-sdk-python/blob/7.0.0.1/LICENSE

It is MIT.

Em sáb, 25 de abr de 2020 13:33, Jc2k <notifications@github.com> escreveu:

**Comment by Jc2k** [comment]
Well that’s exactly why the community is confused about the lack of a pypi release. They say it is MIT just as you do, but the VMware team themselves also say it is not open source enough to go on pypi. They haven’t provided much clarity on what the problem is. Just that it’s a legal one. So maybe only a part of the release is MIT?

**Comment by rmontenegroo** [comment]
I guess I will code some direct access to VMware api.

Em sáb, 25 de abr de 2020 14:49, Jc2k <notifications@github.com> escreveu:

File: vmware_exporter/helpers.py (hunk 1 above)
Comment (rmontenegroo) [reply vmware_exporter/helpers.py]: I am pushing a version where I am replacing vmware SDK for direct requests to vcenter REST API so as to fetch objects tags. So this bit of code won't exist.

File: vmware_exporter/helpers.py (hunk 2 above)
Comment (rmontenegroo) [reply vmware_exporter/helpers.py:23]: I dont know what you mean but in a metric a custom attribute would look something like this:
```
vmware_vm_net_droppedTx_summation{Role="Nginx",ClientID="36310",Configurator="John Doe",Creator="Jane Doe",ApplicatioName="HelloWorld",CreationTicket="2019/000215454",cluster_name="Cluster0018",dc_name="DataCenter02",host_name="dc02cl00018host1111.domain.com",tags="n/a",vm_name="vm-nginx4781"} 0.0
```

File: vmware_exporter/vmware_exporter.py (hunk 3 above)
Comment (rmontenegroo) [reply vmware_exporter/vmware_exporter.py:86]: Oh! Thanks! I thought it would like more synthetic and less places to change if needed.

**Comment by rmontenegroo** [comment]
> Well that’s exactly why the community is confused about the lack of a pypi release. They say it is MIT just as you do, but the VMware team themselves also say it is not open source enough to go on pypi. They haven’t provided much clarity on what the problem is. Just that it’s a legal one. So maybe only a part of the release is MIT?

You're right. Definitely there is something weird. They say there might be a license issue with the code generator they use but it does not seem to be really. They would not share the code on github if it were not "open" enough. And 2 years is quite enough term to at least propose a solution.

Anyways... I've replaced their python sdk by direct request to vsphere rest api. I hope it is a bettter approach.

**Comment by rmontenegroo** [comment]
Hey! I had to make some minor adjustment:

1. I missed the get_env_bool;
2. I replaced vmware sdk by rest api direct requests for tag fetching;
3. I sent an email about this third change. The collector deferred methods, when the result is anything that can be coerced to False, hangs the method. I made some change on that, and I would like you to check.

With those change above, you will need to make a new release, maybe v0.12.1 ? :-P

**Comment by alecorps** [comment]
nice job

**Comment by rmontenegroo** [comment]
I am no git expert. Please check if commits are squashed as requested.

**Comment by rmontenegroo** [comment]
@pryorda Am I missing something? How can I help?

**Comment by pryorda** [comment]
> @pryorda Am I missing something? How can I help?

Sorry about the delay. I'm testing this out now.

**Pull request merged**